### ISSUE TITLE
When all other inputs are 0, right trigger acts as if it's being pressed.

### ISSUE THREAD

**Issue by Mocha-Bean** [kozec/sc-controller#21]
This is my best understanding of the problem, through experimentation in Rocket League and also through testing with [jstest](https://www.archlinux.org/packages/community/x86_64/joyutils/).

If I am not manipulating any other inputs on the controller, then right trigger will act as if it is pressed.  But when I am manipulating another input, it will act properly. This is always reproducible.

https://youtu.be/QbsFHys_DAg

Here's my profile:

```
{
	"buttons" : {
		"A"      : { "action" : "button(Keys.BTN_A)" },
		"B"      : { "action" : "button(Keys.BTN_B)" },
		"X"      : { "action" : "button(Keys.BTN_X)" },
		"Y"      : { "action" : "button(Keys.BTN_Y)" },
		"LB"     : { "action" : "button(Keys.BTN_TL)" },
		"RB"     : { "action" : "button(Keys.BTN_TR)" },
		"BACK"   : { "action" : "button(Keys.BTN_SELECT)" },
		"START"  : { "action" : "button(Keys.BTN_START)" },
		"C"      : { "action" : "button(Keys.BTN_MODE)" },
		"LPAD"   : { "action" : "button(Keys.BTN_THUMBL)" },
		"RPAD"   : { "action" : "button(Keys.BTN_THUMBR)" },
		"LGRIP"  : { "action" : "button(Keys.BTN_A)" },
		"RGRIP"  : { "action" : "button(Keys.BTN_X)" }
	},
	
	"trigger_left":  { "action": "axis(Axes.ABS_Z)" },
	"trigger_right": { "action": "axis(Axes.ABS_RZ)" }
	,
	"gyro": {}, 
	
	"stick" : {
		"X"	: { "action" : "axis(Axes.ABS_X)" },
		"Y"	: { "action" : "raxis(Axes.ABS_Y)" }
	},
	
	"pad_left" : {
		"action" : "dpad(hatup(Axes.ABS_HAT0Y), hatdown(Axes.ABS_HAT0Y), hatleft(Axes.ABS_HAT0X), hatright(Axes.ABS_HAT0X) )"
	},
	
	"pad_right" : {
		"X"	: { "action" : "axis(Axes.ABS_RX)" },
		"Y"	: { "action" : "raxis(Axes.ABS_RY)" },
		"feedback": ["RIGHT", 256]
	}
}
```

**Comment by kozec** [comment]
Ok, three questions:

1. What happens when you do press right trigger?
2. If you switch left and right trigger around, as in
```

    "trigger_left":  { "action": "axis(Axes.ABS_RZ)" },
    "trigger_right": { "action": "axis(Axes.ABS_Z)" }
```
which one acts like this?
3. What version are using?

**Comment by Mocha-Bean** [comment]
**3.** I was on 0.2.3.r46. I just updated to 0.2.3.r61. The problem persists, and nothing about the problem's nature has changed.

**1.** When I _am_ pressing the right trigger, everything works fine. It gives the correct output, based on how far I am pressing it. But once I let go, it assumes the random value (as you can see in the video I provided) around ~16000.

**2.** If I switch the triggers as you suggested, trigger_right (i.e. the right trigger on my steam controller) is still the one with the problem, even though it is now emulating a left trigger.

Also, you know (as I described in the video) how when I press a button (A, X, Y, B, back buttons, start, select, etc.), the right trigger value is corrected back to -32767 on the instances of pressing said buttons and releasing them, but not while holding those buttons? That actually also applies to axis values, but it's harder to reproduce with those because there's so much rapid, minute variation in an axis that it's hard for an axis value to stay the same for a long period of time. Also, a side problem is that when the right trigger value goes back to the random ~16,000 number, all other values being held go back to their default values (0 for buttons and joysticks, -32767 for triggers).

So, to sum it up, the essence of the problem is better described as:

When no values are being actively changed, the right trigger value will become ~16,000 and all other values will become their defaults.

**Comment by kozec** [comment]
Very traditionally, I can't reproduce this no matter what :D

Please, try this:
Turn off sc-controller, kill daemon (`killall scc-daemon`) and try running this script:
```
from scc.controller import SCController

def callback(sc, time, state):
	print "L:", state.ltrig, "R", state.rtrig

SCController(callback).run()
```

It will simply output hardware state of **L**eft and **R**ight trigger. If values there are similarly broken, it may mean that your controller went nuts, but that would be "fixable" by setting limits on input values.

Also, if you have original [ynsta's driver](https://github.com/ynsta/steamcontroller) installed, you can get similar outuput by running `sc-dump.py`.

**Comment by Mocha-Bean** [comment]
Exactly how and where do I run that script?

**Comment by Mocha-Bean** [comment]
Yes, the values when running that script are similarly broken, in the exact same fashion. R is stuck at about 196.

But it can't be a hardware problem; it doesn't occur in native steam. It only occurs with this driver. Perhaps it's a problem with another piece of software that we have different versions of?

Also, as an update to my blanket statement (When no values are being actively changed, the right trigger value will become ~196 and all other values will become their defaults.), the "when no values are being changed" doesn't apply to triggers. If I hold the left trigger down, right trigger will remain correctly at 0 until I entirely let go of the left trigger. The only axis input it applies to is joysticks.

**Comment by kozec** [comment]
> Also, I just tested it: yes, the problem happens in ynsta's driver too.

I thought so... Fun thing is that those numbers really are raw data directly from bus.

But you said you have Steam... Can you check controller fw version? It's displayed only in Big Picture Mode in settings -> controller add/test -> click on controller -> click on 'support' button. Even screenshot from that screen would be nice.
And, by the way, if that screen displays same behavior, it explains everythin :P

And last question, what distro are you using? I'll try to run it in Virtualbox and reproduce this...

**Comment by Mocha-Bean** [comment]
![](http://i.imgur.com/hOOXunj.png)

I ran the test, and it completed without errors. (And, yes, the issue also does not present itself when using the native Steam driver.) I specifically paid attention to the right trigger test to make sure that it started on 0 (unlike how it would in ynsta's driver or yours).

**Comment by kozec** [comment]
Sooo... you are using exactly same FW versions, exactly same kernel and almost same distro as me, but your controller behaves differently, unless it talks directly to Steam.

Do you have any idea why is radio firmware displayed as 0x0 ?

**Comment by kozec** [comment]
... oh gods :D

Are you using controller connected to USB cable instead of wireless dongle by any chance?

**Comment by Mocha-Bean** [comment]
Yes.

**Comment by kozec** [comment]
Well, it looks like Steam Controller connected to USB behaves *almost* exactly same as dongle with controller connected wirelessly. You probably already guessed what is that small difference.

I'll check if there is any sane way how to filter out those random numbers.

**Comment by Mocha-Bean** [comment]
Oh, yeah, I just tested it wirelessly and it works perfectly.
